python3Packages.gradio: init at 3.20.1

pbsds opened the pull request:
###### Description of changes

This PR adds gradio by @gradio-app, a web server framework for making AI (anything really) demos.

Original PR is #187032, which was closed due to a mass-ping resulting from a poorly thought out rebase onto staging.

Close #235184

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Pinging reviewers from prior pr: 
@happysalada @Janik-Haag @VergeDX @SuperSandro2000

pbsds committed:
python3Packages.ffmpy: init at 0.3.1

Co-authored-by: Yt <happysalada@tuta.io>

pbsds committed:
python3Packages.gradio: init at 3.20.1

Co-authored-by: Yt <happysalada@tuta.io>

pbsds committed:
python3Packages.markdown-it-py: fix linkify extra

linkify-it-py is on version 2.0.0, while markdown-it-py[linkify] depends on linkify-it-py~=1.0

pbsds commented:
@ofborg build python3Packages.gradio

happysalada commented:
Have you seen usage of nix-update-script ?
It will generate automatic prs with updates of your packages. It will reduce the cost of maintenance quite a bit in my experience.
If you search for it in the repo there are plenty if exemples.
Sometimes it takes 1-2 days to generate an update but thats still better than having to do it by hand everytime.

happysalada commented:
@mweinelt the ci fails on aarch64-linux, the logs don't seem to say much. This could be due to a timeout. Is this of any concern to you ?

happysalada commented:
@pbsds this looks good to me, do you intend to do any more work on this ?

happysalada commented:
@pbsds I'm waiting for this to package FastChat, let me know if I can do anything to help.

mweinelt commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,34 @@
+    sha256 = "dXWRWB7uJbSlCsn/ubWANaJ5RTPbR+BRL1P7LXtvmtw=";
+  };
+
+  propagatedBuildInputs = [
+    ffmpeg
+  ];
```
https://github.com/Ch00k/ffmpy/blob/master/ffmpy.py#L14

Instead of propagating ffmpeg, replace the executable path here with the absolute path to the ffmpeg executable

mweinelt commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,34 @@
+buildPythonPackage rec {
+  pname = "ffmpy";
+  version = "0.3.0";
+
```
Add an explicit format.

```suggestion
  format = "setuptools";
```

mweinelt commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,34 @@
+  ];
+
+  pythonImportsCheck = [ "ffmpy" ];
+
```
```suggestion

  doCheck = false; # no tests in sdist, no tags on github

```

Disable the tests, if there are none. The situation is somewhat common, so we can probably skip the bigger comment above src.

pbsds replied:
https://github.com/Ch00k/ffmpy/issues/60 resolved, i switched to fetchFromGitHub and now run pytest

mweinelt commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+    typing-extensions
+  ] ++ markdown-it-py.optional-dependencies.linkify;
+
+  nativeBuildInputs = [
+    hatchling
+    hatch-requirements-txt
+    hatch-fancy-pypi-readme
+  ];
```
Move these above `propagatedBuildInputs`.

mweinelt commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+    hatch-fancy-pypi-readme
+  ];
+
+  checkInputs = [
```
```suggestion
  nativeCheckInputs = [
```

mweinelt reviewed (changes requested):
The test phase of gradio looks annyoing to maintain.

happysalada commented:
@pbsds hey, it looks like you are a little busy, would you mind If I took over ?

pbsds commented:
> @pbsds hey, it looks like you are a little busy, would you mind If I took over ?

My computer and network access this summer is spotty. I added you to my nixpkgs repo if you're interested in joint maintainership :)

@ofborg build python3Packages.gradio

happysalada commented:
This looks good to me.
Im happy to be pinged in case there are problems discovered during the staging PR.
@mweinelt might have more comments, so waiting a bit more.

mweinelt commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,52 @@
+
+  pythonImportsCheck = [ "ffmpy" ];
+
+  nativeCheckInputs = [ pytestCheckHook go ];
```
```suggestion
  nativeCheckInputs = [
    pytestCheckHook
    go
  ];
```

Please expand lists to one item per line.

mweinelt commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,52 @@
+    substituteInPlace ffmpy.py \
+      --replace 'executable="ffmpeg",' 'executable="${ffmpeg}/bin/ffmpeg",'
+
+    # fix tests assuming "ffmpeg"
+    for fname in tests/*.py; do
+      echo 'FFmpeg.__init__.__defaults__ = ("ffmpeg", *FFmpeg.__init__.__defaults__[1:])' >>"$fname"
+    done
```
We could just pass `ffmpeg` into `nativeCheckInputs`, no?

happysalada replied:
I just tested this different scenarios

removing the `# fix tests ` code block (including the for loop) and adding ffmpeg to nativecheckinputs doesn't work unfortunately

The failures are of type
```
 AssertionError: assert '/nix/store/f...-exit-code 42' == 'ffmpeg --std...-exit-code 42'
ffmpy> E         - ffmpeg --stdin none --stdout none --stderr multiline --exit-code 42
ffmpy> E         + /nix/store/fy4nryi6sg44iy0ipq1j46vq8v9hq5v3-ffmpeg-5.1.3-bin/bin/ffmpeg --stdin none --stdout none --stderr multiline --exit-code 42
```

strangely enough, the tests pass when removing the following code
```
    # default to store ffmpeg
    substituteInPlace ffmpy.py \
      --replace 'executable="ffmpeg",' 'executable="${ffmpeg}/bin/ffmpeg",'
```

since I'm not sure how to properly test ffmpy, I suggest to leave as is.

pbsds replied:
The tests test a mock that do not behave like ffmpeg. If we default to in-store ffmpeg they fail.

happysalada replied:
I added that exact comment to clarify the situation.

mweinelt commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,52 @@
+  postPatch = ''
+    # default to store ffmpeg
+    substituteInPlace ffmpy.py \
+      --replace 'executable="ffmpeg",' 'executable="${ffmpeg}/bin/ffmpeg",'
```
```suggestion
      --replace 'executable="ffmpeg",' 'executable="${ffmpeg-headless}/bin/ffmpeg",'
```

Please check whether the headless variant of ffmpeg is suitable. I would imagine it could be.

happysalada replied:
confirmed it's working.

mweinelt commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+      from _pytest.runner import pytest_runtest_makereport as orig_pytest_runtest_makereport
+      import urllib, urllib3, httpx, requests, websockets
+
+      class NixNetworkAccessDeniedError(BaseException): pass
+      def deny_network_access(*a, **kw):
+        raise NixNetworkAccessDeniedError
+
+      def iterate_exc_chain(exc: Exception):
+        yield exc
+        if getattr(exc, "__context__", None) is not None:
+          yield from iterate_exc_chain(exc.__context__)
+        if getattr(exc, "__cause__", None) is not None:
+          yield from iterate_exc_chain(exc.__cause__)
+
+      httpx.Request = \
+      httpx.AsyncClient.get = \
+      httpx.AsyncClient.head = \
+      requests.head = \
+      requests.get = \
+      requests.post = \
+      urllib.request.urlopen = \
+      urllib.request.Request = \
+      urllib3.connection.HTTPSConnection._new_conn = \
+      websockets.connect = \
+        deny_network_access
+
+      def pytest_runtest_makereport(item, call):
+        tr = orig_pytest_runtest_makereport(item, call)
+        if call.excinfo is not None:
+          for exc in iterate_exc_chain(call.excinfo.value):
+            if isinstance(exc, NixNetworkAccessDeniedError):
+              tr.outcome, tr.wasxfail = 'skipped', "reason: Requires network access."
+            if isinstance(exc, FileNotFoundError):
+              tr.outcome, tr.wasxfail = 'skipped', "reason: Pypi dist bad."
+        return tr
+    '';
+  in ''
+    export HOME=$TMPDIR
+    cat ${conftestSkipNetworkErrors} >> test/conftest.py
+  '';
```
Is some of that upstreamable? Like a `network` marker? As I mentioned, this looks annoying to maintain.

happysalada replied:
IMO this is too related to nix to matter upstream.
How about we keep this as is, and if for some reason this doesn't work anymore, we just disable the tests until someone has time to look into this.
Since there are quite a lot of packages depending on gradio, would you rather 
- disable the tests until we confirm with upstream if they are interested in this or not.
- leave this as is for now and disable the tests if there is a failure related to this.

SuperSandro2000 replied:
> IMO this is too related to nix to matter upstream.

Having an extra marker to skip network tests is something not nix specific, many distros build their packages without networking. It can also be a great way to skip tests if no network is available or only very slow and flaky.

pbsds replied:
In the prior pr I discussed generalizing this into a pytestCheckHook-esqe hook. The sentiment then was first pushing through this pr

SuperSandro2000 replied:
If the usecase comes up, we can always move the implementation into a hook but I am not fully convinced that this is necessary here.

What's the problem with upstreaming this and adding a marker to skip network tests? We just have a build environment without internet access but we would still like to run as many tests as possible.

pbsds replied:
I don't trust the gradio devs, who tend to move fast and loose, to support this use case. My reason is that i did [this pr](https://github.com/gradio-app/gradio/pull/1706), to fix the missing test files on pypi. They they promptly broke it again without a few weeks.

I have bumped the version of the prior gradio pr about 3 times over the past year, and each time the list of tests to disable needed change. This proved quite time consuming  to maintain (hence that debug helper comment). Network and non-network tests are not grouped in any meaningful sense, so `disabledTestPaths` proved too aggressive while `disabledTests` proved too surgical and unmaintainable.

The conftest network xfail approach came upon me in a dream as an easily generalizable drop-in way to run as many tests as possible, without maintaining a `disabledTests` list, nor relying on upstream devs to support us poor distro maintainers. It is my brain child and I will defend it with passion :P.

To argue in favor of generalizing this xfail approach and a future check hook, here is a very coarse estimate of the number of packages which disable network tests, and likely a lot more, with `disableTestPaths`:

```shell
$ rg disabledTestPaths pkgs/development/python-modules/ -l | xargs rg network | wc -l
137
```

The conftest network xfail approach will for these likely improve testing coverage by a lot, and I believe such a hook will even remove a lot of `doChecks = false;` found in throughout our python package set.

happysalada commented on pkgs/top-level/python-packages.nix:
```diff
@@ -3596,6 +3596,10 @@ self: super: with self; {
   ffmpeg-progress-yield = callPackage ../development/python-modules/ffmpeg-progress-yield { };
 
+  ffmpy = callPackage ../development/python-modules/ffmpy {
+    inherit (pkgs) ffmpeg;
```
```suggestion
    ffmpeg = pkgs.ffmpeg-headless;
```

happysalada commented:
@mweinelt ready for another round of review.
I've added two commits that will later be squashed. I didn't want to replace the commits so that the OP would loose attribution.
let me know if anything.

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+    pytestCheckHook
+    pytest-asyncio
+    mlflow
+    #cometml # FIXME: enable once comet-ml is packaged
```
```suggestion
    #comet-ml # FIXME: enable once comet-ml is packaged
```

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+
+  postPatch = ''
+    substituteInPlace requirements.txt \
+      --replace \
+        "mdit-py-plugins<=0.3.3" \
+        "mdit-py-plugins"
```
```suggestion
      --replace "mdit-py-plugins<=0.3.3" "mdit-py-plugins"
```

happysalada replied:
Could we keep the formatting as is ? The multi-line makes it easy to verify there are no typos.

Janik-Haag replied:
isn't that what `pythonRelaxDeps = [ ];` is for?

pbsds replied:
It works now? That is news to me

Janik-Haag replied:
It worked for me months ago, but you have to be careful with the capitalization and `-` and so on in the strings and can't just use the pname of the dep.

happysalada replied:
I didn't know this hook was working with pyproject.
I tested this overnight however and it does work. I used exactly `mdit-py-plugins` as the dependency to relax.

Janik-Haag replied:
Then I suggest we switch to this since it's a "cleaner" way :D

happysalada replied:
That works for me . @pbsds ill let you make the change when you come back

pbsds replied:
Doesn't work on my end

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+, boto3
+}:
+
+let
+  # breaks way too frquently
+  shapWorks = (builtins.tryEval shap.outPath).success;
+in
+
```
that's rather hacky, I would like to avoid this

pbsds replied:
Shape was the reason the prior pr stalled

SuperSandro2000 replied:
We could just disable the tests all the time to avoid this. I am not even sure if tryEval has implications.

happysalada replied:
I've removed those lines completely and disabled by default the problematic test.

mweinelt commented on pkgs/development/python-modules/markdown-it-py/default.nix:
```diff
@@ -36,6 +36,14 @@ buildPythonPackage rec {
+  # fix downstrem usage of markdown-it-py[linkify]
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace \
+        'linkify = ["linkify-it-py~=1.0"]' \
+        'linkify = ["linkify-it-py"]'
```
Disagree, multiline makes it much better readable.

Janik-Haag replied:
Same as above:
isn't that what `pythonRelaxDeps = [ ];` is for?

happysalada commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+    # Flaky (very sensitive to dep version)
+    "test_in_interface_as_output"
+    "test_should_warn_url_not_having_version"
+  ]
+  ++ lib.optionals (!shapWorks) [
```
```suggestion
# shap breaking often makes this test flakey
```

happysalada replied:
@pbsds isn't the boolean the opposite here ?
Maybe I'm understanding wrong here, but if shap works include the test, if it doesn't , then don't include it.
I added a suggestion so as to disable the test always to get rid completely of the shapworks evaluation.

pbsds replied:
Yes, sorry

happysalada commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,203 @@
+    # Flaky (very sensitive to dep version)
+    "test_in_interface_as_output"
+    "test_should_warn_url_not_having_version"
+# shap breaking often makes this test flakey
```
```suggestion
    # shap breaking often makes this test flakey
```

happysalada commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,203 @@
+, boto3
+}:
+
+let
+  # breaks way too frquently
+  shapWorks = (builtins.tryEval shap.outPath).success;
+in
+
```
```suggestion
```

happysalada commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,55 @@
+  postPatch = ''
+    # default to store ffmpeg
+    substituteInPlace ffmpy.py \
+      --replace 'executable="ffmpeg",' 'executable="${ffmpeg}/bin/ffmpeg",'
```
```suggestion
      --replace 'executable="ffmpeg",' 'executable="${ffmpeg-headless}/bin/ffmpeg",'
```

happysalada commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,198 @@
+    ffmpeg
+    vega_datasets
+    boto3
+    # shap is needed as well but breaks too often
```
```suggestion
    # shap is needed as well but breaks too often
  ];
```

happysalada commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,204 @@
+    # makes pytest freeze 50% of the time
+    "test/test_interfaces.py"
+  ];
+  #pytestFlagsArray = [ "-x" "-W" "ignore" ]; # uncomment for debugging help
```
if we start getting failures on the package, I would say let's enable it. For now it's best to leave it disabled I think.

happysalada commented:
@SuperSandro2000 I think I've addressed all of your comments, but let me know if there is something that you think should be addressed differently.

I've also just tested to make sure this actually works.

SuperSandro2000 reviewed (commented):
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

pbsds commented:
I think this pr, with a few squashes, looks good! Thank you so much @happysalada for being such a champ in pushing this though.

Once some time when I'm back in civilization I will do a pr to bump gradio to handle the cyclical depdendency on gradio-client introduced in the latest version, and try to hoist the conftest network xfail code to a drop-in check hook.

happysalada commented:
@pbsds no thank you for making this PR in the first place ! This has been a long PR and you stayed with it, all the more kudos to you.

Enjoy your time outside of civilisation ! (And dont feel any rush to come back because of this, this can definitely wait)

pbsds commented:
I squashed your commits, adding you as co-author. This now builds on master, and I leave checking the 2k+ affected packages to the staging gods.

happysalada commented:
This looks good to me, waiting another 24h just in case someone else wants to have a look

Janik-Haag commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,200 @@
+buildPythonPackage rec {
+  pname = "gradio";
+  version = "3.20.1";
+  format = "pyproject";
+  disabled = pythonOlder "3.7";
```
```suggestion
  format = "pyproject";

  disabled = pythonOlder "3.7";
```

Janik-Haag commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,200 @@
+  # Add a pytest hook skipping tests that access network, marking them as "Expected fail" (xfail).
+  # We additionally xfail FileNotFoundError, since the gradio devs often fail to upload test assets to pypi.
+  preCheck = let
+    conftestSkipNetworkErrors = writeText "conftest-network-xfail.py" ''
```
I think this should be a patch since it gets appended to a existing source file.

so probably something like this. 
```
# Add a pytest hook skipping tests that access network, marking them as "Expected fail" (xfail).
# We additionally xfail FileNotFoundError, since the gradio devs often fail to upload test assets to pypi.
patches = [
  ./conftest-skip-network-errors.patch
];
```

there is currently a WIP pr describing how we should handle patches, it's obviously not a rule since it's not merged yet but might give you pointers in the right direction. #246766

pbsds replied:
The plan is to generalize this pytest hook into a generic mkDerivation hook, i.e. `pytestXfailNetworkHook`, which works for any pytest-based project. It was written this way to make it work with any `conftest.py` source, even in cases where there are no `conftest.py` source already. Turning it into a patch would make this approach far less general and indicate that the code only applies only to gradio, which is does not.

If preferable I can move the python code into a separate file for this PR, but turning it into a patch would only make it less generic and turn it into a maintenance burden.

Janik-Haag replied:
Okay I guess it's fine then, thank you for the explanation.

Janik-Haag reviewed (commented):
Except for this nitpicks this should be good to go.
I'll have to check later why the tests are failing.

happysalada commented:
@Janik-Haag there are 2 potential improvements (1 formatting and one that could be discussed potentially later). Would you mind if we merged and kept those further improvements for a subsequent PR. The current patch is still in a "let's try it and we will see if we like it" state, so I would say it would be a good candidate to revisit this in 6 months.
gradio is a dependency used by many machine learning tools, there is value in having it available earlier than later.
Feel free to let me know if you disagree, no problem at all.

Janik-Haag commented:
I would prefer to do it right the first time since it's not a lot of work to do these changes and we currently can't merge it anyways since the builds are failing and it doesn't seem to be because of a timeout.

pbsds commented:
Fixed the formatting issue, and move the python code to a separate file. Now you get nice colors in your editor.

As (elaborate here)[https://github.com/NixOS/nixpkgs/pull/241558#discussion_r1285123038] i opt not to turn the python code into a patch.

Of course the build fails, it's staging and the closure is ginormous. When cherry-picked on master it builds fine.

happysalada commented:
I wanted to make sure the build passed. After testing on a x86_64-linux machine I have a consistent failure with just one of the tests.
here is the stacktrace

```
gradio> test/test_examples.py xxxx.xxxxx.xx.x.....xx                             [ 47%]
gradio> test/test_external.py xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx..xxxxxx......... [ 58%]
gradio> ...xxx                                                                   [ 60%]
gradio> test/test_flagging.py xx.x.xx......                                      [ 63%]
gradio> test/test_interpretation.py .........                                    [ 65%]
gradio> test/test_mix.py .x..x                                                   [ 66%]
gradio> test/test_networking.py ..xx.x                                           [ 67%]
gradio> test/test_pipelines.py x                                                 [ 68%]
gradio> test/test_processing_utils.py .x.x..............x..........x             [ 75%]
gradio> test/test_queueing.py ...............xxx.x....                           [ 80%]
gradio> test/test_reload.py .                                                    [ 81%]
gradio> test/test_routes.py xxxxxxxxxxxxxx.xx...xxxxxxxx                         [ 87%]
gradio> test/test_tunneling.py x                                                 [ 87%]
gradio> test/test_utils.py .xxx....x.......x........xxxxxxxxxxx...xx...........  [100%]
gradio>
gradio> =================================== FAILURES ===================================
gradio> __________________________ TestJSON.test_in_interface __________________________
gradio>
gradio> x = array(['MM'], dtype=object)
gradio>
gradio>     def _ensure_numeric(x):
gradio>         if isinstance(x, np.ndarray):
gradio>             if is_integer_dtype(x) or is_bool_dtype(x):
gradio>                 x = x.astype(np.float64)
gradio>             elif is_object_dtype(x):
gradio>                 try:
gradio> >                   x = x.astype(np.complex128)
gradio> E                   ValueError: complex() arg is a malformed string
gradio>
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>
gradio> During handling of the above exception, another exception occurred:
gradio>
gradio> x = array(['MM'], dtype=object)
gradio>
gradio>     def _ensure_numeric(x):
gradio>         if isinstance(x, np.ndarray):
gradio>             if is_integer_dtype(x) or is_bool_dtype(x):
gradio>                 x = x.astype(np.float64)
gradio>             elif is_object_dtype(x):
gradio>                 try:
gradio>                     x = x.astype(np.complex128)
gradio>                 except (TypeError, ValueError):
gradio>                     try:
gradio> >                       x = x.astype(np.float64)
gradio> E                       ValueError: could not convert string to float: 'MM'
gradio>
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>
gradio> The above exception was the direct cause of the following exception:
gradio>
gradio> self = <test.test_components.TestJSON object at 0x7fff767a4670>
gradio>
gradio>     @pytest.mark.asyncio
gradio>     async def test_in_interface(self):
gradio>         """
gradio>         Interface, process
gradio>         """
gradio>
gradio>         def get_avg_age_per_gender(data):
gradio>             print(data)
gradio>             return {
gradio>                 "M": int(data[data["gender"] == "M"].mean()),
gradio>                 "F": int(data[data["gender"] == "F"].mean()),
gradio>                 "O": int(data[data["gender"] == "O"].mean()),
gradio>             }
gradio>
gradio>         iface = gr.Interface(
gradio>             get_avg_age_per_gender,
gradio>             gr.Dataframe(headers=["gender", "age"]),
gradio>             "json",
gradio>         )
gradio>         y_data = [
gradio>             ["M", 30],
gradio>             ["F", 20],
gradio>             ["M", 40],
gradio>             ["O", 20],
gradio>             ["F", 30],
gradio>         ]
gradio> >       assert (
gradio>             await iface.process_api(
gradio>                 0, [{"data": y_data, "headers": ["gender", "age"]}], state={}
gradio>             )
gradio>         )["data"][0] == {
gradio>             "M": 35,
gradio>             "F": 25,
gradio>             "O": 20,
gradio>         }
gradio>
gradio> test/test_components.py:1746:
gradio> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
gradio> gradio/blocks.py:1032: in process_api
gradio>     result = await self.call_function(
gradio> gradio/blocks.py:844: in call_function
gradio>     prediction = await anyio.to_thread.run_sync(
gradio> /nix/store/m8ix667n94csdc81m5aq55kfypc1bn3c-python3.10-anyio-3.7.0/lib/python3.1
gradio>     return await get_asynclib().run_sync_in_worker_thread(
gradio> /nix/store/m8ix667n94csdc81m5aq55kfypc1bn3c-python3.10-anyio-3.7.0/lib/python3.1
gradio>     return await future
gradio> /nix/store/m8ix667n94csdc81m5aq55kfypc1bn3c-python3.10-anyio-3.7.0/lib/python3.1
gradio>     result = context.run(func, *args)
gradio> test/test_components.py:1729: in get_avg_age_per_gender
gradio>     "M": int(data[data["gender"] == "M"].mean()),
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return NDFrame.mean(self, axis, skipna, numeric_only, **kwargs)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return self._stat_function(
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return self._reduce(
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     res = df._mgr.reduce(blk_func)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     nbs = blk.reduce(func)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     result = func(self.values)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return op(values, axis=axis, skipna=skipna, **kwds)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return f(*args, **kwargs)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     result = alt(values, axis=axis, skipna=skipna, **kwds)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     result = func(values, axis=axis, skipna=skipna, mask=mask, **kwargs)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     the_sum = _ensure_numeric(values.sum(axis, dtype=dtype_sum))
gradio> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
gradio>
gradio> x = array(['MM'], dtype=object)
gradio>
gradio>     def _ensure_numeric(x):
gradio>         if isinstance(x, np.ndarray):
gradio>             if is_integer_dtype(x) or is_bool_dtype(x):
gradio>                 x = x.astype(np.float64)
gradio>             elif is_object_dtype(x):
gradio>                 try:
gradio>                     x = x.astype(np.complex128)
gradio>                 except (TypeError, ValueError):
gradio>                     try:
gradio>                         x = x.astype(np.float64)
gradio>                     except ValueError as err:
gradio>                         # GH#29941 we get here with object arrays containing strs
gradio> >                       raise TypeError(f"Could not convert {x} to numeric") from err
gradio> E                       TypeError: Could not convert ['MM'] to numeric
gradio>
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.10/site-packages/pandas/core/nanops.py:1686: TypeError
gradio> ----------------------------- Captured stdout call -----------------------------
gradio>   gender  age
gradio> 0      M   30
gradio> 1      F   20
gradio> 2      M   40
gradio> 3      O   20
gradio> 4      F   30
```

Nothing jumps at me in particular, so I would advise just disabling the test. Let me know if you disagree.

(in the process of testing I discovered several failures with upstream dependencies that I have reported. Note where related to the changes made here.)

Janik-Haag commented:
Are you sure the package even works if there are so many test failures?

happysalada commented:
First step is merging i think. Trying to rebuild all staging everytime to test us just not very efficient. First we make sure nothing breaks, then we should be able toake changes that impact much less things. I have locally packaged fast chat, ill open a PR as soon as this hits master.

Janik-Haag commented:
You can just do `git switch master` and then cherry pick the commits like this: `git cherry-pick fdfcecedb72326c19772b4720a2e9c6df46347db 4324abf898a0ff5a14fe4b740f81f97ddc2e34a8 90500b167084b87a385c68f29de128418bc2da35` if you checked out this branch locally with `gh pr checkout 241558`. and then you can test locally without problem

pbsds commented:
I'm unable to reproduce your error. I also tried python 3.11, but that was blocked by tensorflow being marked broken on master. I'm disabling `test_in_interface` due to being flaky.

> Are you sure the package even works if there are so many test failures?

The gradio devs tend to move fast and break things, and the functionality they provide is incredibly simple but pimped with lots of fancy webdev. The gradio test suite is mostly integration testing with all the major ML frameworks, hence the ginormous check closure.

Perhaps the future is `doCheck = false`, but as of now i consider this testing coverage a major success 
![image](https://github.com/NixOS/nixpkgs/assets/140964/5743cc58-badc-47d5-9052-197dfb824c73)

pbsds commented:
> You can just do `git switch master` and then cherry pick the commits like this: `git cherry-pick fdfcecedb72326c19772b4720a2e9c6df46347db 4324abf898a0ff5a14fe4b740f81f97ddc2e34a8 90500b167084b87a385c68f29de128418bc2da35` if you checked out this branch locally with `gh pr checkout 241558`. and then you can test locally without problem

I use this one-liner, assuming this per is checked out in the worktree `../gradio-init`: 

```
git checkout master;  (cd ../gradio-init/; git log --oneline) | head -n3 | cut -d\  -f1 | tac | xe git cherry-pick
```

I'm surprised how nixpkgs-review does not have a cherry-pick or rebase strategy for staging PRs. (https://github.com/Mic92/nixpkgs-review/issues/287)

happysalada commented:
@Janik-Haag have your concerns been addressed ?


Regarding my test failure, ive had other test failures on staging that couldbt be reproduced. I have had one on thrift, that was tested on another machine and couldnt be reproduced. So the one i had could be just a fluke.

Janik-Haag commented:
I currently am in a train so I have whonky internet, I'll try to check the package in the next ~8 hours once I have a stable connection again. But I really don't want to merge anything broken.

happysalada commented:
At this point, i dont think this paclage is broken, i think there could be flakiness in the tests or some potential problems with downstream dependents, but am not sure how we much more we could test.

Happy to wait one more day if you think we can potentially get more information.

pbsds commented:
I rebased onto the recent merge base between master and staging, and now it builds. So far I've waited one hour for the ofborg eval to be done. I'll marking the PR ready again once the eval is done, just for safety.

Janik-Haag commented:
It now builds fine but when I try to run the binary it throws this error:
```
./result/bin/gradio
Traceback (most recent call last):
  File "/nix/store/3nap39bl8f7wbrbzfkl1dbj15lnab8aj-python3.10-gradio-3.20.1/bin/.gradio-wrapped", line 6, in <module>
    from gradio.reload import run_in_reload_mode
  File "/nix/store/3nap39bl8f7wbrbzfkl1dbj15lnab8aj-python3.10-gradio-3.20.1/lib/python3.10/site-packages/gradio/__init__.py", line 3, in <module>
    import gradio.components as components
  File "/nix/store/3nap39bl8f7wbrbzfkl1dbj15lnab8aj-python3.10-gradio-3.20.1/lib/python3.10/site-packages/gradio/components.py", line 33, in <module>
    from gradio.blocks import Block, BlockContext
  File "/nix/store/3nap39bl8f7wbrbzfkl1dbj15lnab8aj-python3.10-gradio-3.20.1/lib/python3.10/site-packages/gradio/blocks.py", line 24, in <module>
    from gradio import (
  File "/nix/store/3nap39bl8f7wbrbzfkl1dbj15lnab8aj-python3.10-gradio-3.20.1/lib/python3.10/site-packages/gradio/networking.py", line 17, in <module>
    from gradio.routes import App
  File "/nix/store/3nap39bl8f7wbrbzfkl1dbj15lnab8aj-python3.10-gradio-3.20.1/lib/python3.10/site-packages/gradio/routes.py", line 25, in <module>
    import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'
```

pbsds commented:
Great catch! I fixed it, and also added a `postInstallCheck` testing for this.
I'm surprised the pythonImportsCheck runs in the python stdenv.

I hosted a gradio service last year, so it didn't strike me to check the shell entrypoint as I kept bumping the version of the prior PR.

Janik-Haag reviewed (approved):
LGTM now, sorry for all the trouble but I want to avoid merging broken stuff.

happysalada commented:
@Janik-Haag on the contrary, thank you for your thoroughness.

@pbsds thanks for sticking with this PR for so long!

Magellol commented:
Unfortunately, It doesn't look like I can build this on `aarch64-darwin` :/ — Anyone has an idea? I'm not super knowledgeable with the ins and outs of python.

```
error: builder for '/nix/store/j5m87whdmc1ryjy21l99r0rinf3rzmlz-python3.11-gradio-client-0.10.1.drv' failed with exit code 4;
       last 10 log lines:
       > /nix/store/9nasxys60wirmhqs07akfb68n83pip6x-python3.11-gradio-sans-client-4.20.1/lib/python3.11/site-packages/gradio/routes.py:81: in <module>
       >     from gradio.server_messages import (
       > /nix/store/9nasxys60wirmhqs07akfb68n83pip6x-python3.11-gradio-sans-client-4.20.1/lib/python3.11/site-packages/gradio/server_messages.py:64: in <module>
       >     class UnexpectedErrorMessage(BaseModel):
       > /nix/store/9nasxys60wirmhqs07akfb68n83pip6x-python3.11-gradio-sans-client-4.20.1/lib/python3.11/site-packages/gradio/server_messages.py:65: in UnexpectedErrorMessage
       >     msg: Literal[ServerMessage.unexpected_error] = ServerMessage.unexpected_error
       > /nix/store/c05vbvkjxarxkws9zkwrcwrzlsx9nd68-python3-3.11.8/lib/python3.11/enum.py:784: in __getattr__
       >     raise AttributeError(name) from None
       > E   AttributeError: unexpected_error
       > /nix/store/gamm276nsv0l0qm71qhywpz833niaxkx-stdenv-darwin/setup: line 1587: pop_var_context: head of shell_variables not a function context
```
